### PR TITLE
build: move to new protobuf library

### PR DESCRIPTION
**Pull request by DStrand1**
This PR upgrades Telegraf to use the new `google.golang.org/protobuf` library over the deprecated `github.com/golang/protobuf` and abandoned `github.com/gogo/protobuf`. This matches similar work performed in [IDPE](https://github.com/influxdata/idpe/pull/11587), [InfluxDB](https://github.com/influxdata/influxdb/pull/22606), [Chronograf](https://github.com/influxdata/chronograf/pull/5824), [Kapacitor](https://github.com/influxdata/kapacitor/pull/2630), and [InfluxCLI](https://github.com/influxdata/influx-cli/pull/308).

### `jti_openconfig_telemetry` input plugin
The JTI Openconfig plugin was using the `golang` protobuf, and I simply regenerated the existing `.proto` files with the new protobuf library.

### `stackdriver` input/output plugin
Stackdriver was using the `golang` protobuf, and was primarily using it for the known types, which exist in the new Protobuf in an API-compatible way, so these were replaced with the new references.

### `riemann_listener` input plugin
The Riemann Listener plugin was using the `gogo` protobuf, and I resolved this by copying the riemann repo's `.proto` file and generating a new `.pb.go` file using the new protobuf library.

### `cisco_telemetry_mdt` input plugin
The Cisco Telemetry plugin was using a library of pre-generated `.pb.go` files, which lived under `github.com/cisco-ie/nx-telemetry-proto`. Instead of depending on this library, which was using the `golang` protobuf, I instead copied in the `.proto` files from the [original repo](https://github.com/CiscoDevNet/nx-telemetry-proto) and generated new files using the new protobuf library.

### `prometheus` serializer plugin
The Prometheus plugin was using the `gogo` protobuf, however Prometheus generates their `.pb.go` files with `Marshal()` and `Unmarshal()` functions on the proto structs, so Proto was removed from this in favor of directly calling these functions.

The end result of this PR is depending on 3 less libraries (`gogo/protobuf`, `golang/protobuf`, and `cisco-ie/nx-telemetry-proto`), keeping Protobuf consistent across the repo, and avoiding future issues that arise from the deprecated/abandoned repos.

### CODE REVIEW

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:27]: Why do you need to move these to telegraf. I would feel better if we are using the upstream versions instead of copying those into telegraf adding a maintenance burden...

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:66]: Can you elaborate a bit on _why_ it is needed? It was not there previously and I don't see how it is used in the code...

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_util.go
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_util.go:7]: Same, why do we need to move those to telegraf?

File: plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry_test.go
Comment (srebhan) [inline plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry_test.go:50]: Why is this needed?

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (srebhan) [inline plugins/inputs/riemann_listener/riemann_listener.go:24]: Same, why do we need to move those to telegraf?

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (srebhan) [inline plugins/inputs/riemann_listener/riemann_listener.go:184]: Is there a reason you change the spelling? The diff would be much smaller if you change the import.

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (srebhan) [inline plugins/inputs/riemann_listener/riemann_listener.go:398]: Please do not export this function.

File: plugins/serializers/prometheusremotewrite/prometheusremotewrite.go
Comment (srebhan) [inline plugins/serializers/prometheusremotewrite/prometheusremotewrite.go:238]: Please split this into two lines for readablility.

**Review by srebhan — COMMENTED** [review]
Thank you very much @DStrand1 for this PR! It's very much appreciated. My only concern, beside some smaller comments, is moving the protocol buffer files into Telegraf instead of importing them from their upstream location. Can you please either revert this to use the upstream version or elaborate on why this is needed?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (DStrand1) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:66]: `protoc-gen-go-grpc` requires that this be embedded in any struct that uses the proto-gRPC generated types, as it is meant to provide easy forward-compatibility if the generator changes, so that implemented server code doesn't break.

More realistically though, there is an inaccessible member function `mustEmbedUnimplementedGRPCMdtDialoutServer()` created as part of the generated code, that without this embed in the struct, calling `dialout.RegisterGRPCMdtDialoutServer()` on line 171 would give an error. Since it is inaccessible, embedding this is the best way to resolve the issue, and is recommended in a generated code comment (from `dialout_grpc.pb.go#71-73`):
```
// UnimplementedGRPCMdtDialoutServer must be embedded to have forward compatible implementations.
type UnimplementedGRPCMdtDialoutServer struct {
}
```

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (DStrand1) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:27]: I will see if it is possible to generate the `.pb.go` files from vendored `.proto` files instead of copied ones, though these `.pb.go` files will have to be held onto since the repos we are referencing either do not have generated forms of the files, or provide generated forms using outdated libraries

File: plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry_test.go
Comment (DStrand1) [reply plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry_test.go:50]: Same as above

**Comment by DStrand1** [comment]
With regards to copying-in `.proto` files:

**Cisco_telemetry_mdt**: The two options for these `.proto` files are either to copy them in directly and generate our own `.pb.go` files (which we are doing in this PR), or to use the forked repo which provides generated `.pb.go` files (which we were doing before). However, those generated files are using one of the older Protobuf libraries, so they cannot be used any longer. The home repo of these proto files, https://github.com/CiscoDevNet/nx-telemetry-proto, has no packages and nothing real to vendor, so pulling these files in seems easiest to me. They are unlikely to change in impactful ways, as that would break compatibility with previous wire format. It may be possible to still grab this package and generate `.pb.go` files from it similar to below, but I haven't been able to get it working yet.

**Riemann_listener**: This one I was able to change the `go:generate` directive to use the vendored `.proto` file successfully. We may still have to update the generated file if the `.proto` file changes, but it is as easy as re-running `go generate ./...`

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (srebhan) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:66]: Understood. Can you please summarize this as a comment in the code as otherwise someone will remove this and stuff starts to break. :-)

**Comment by srebhan** [comment]
AFAICS there is an open pull-request with this origin https://github.com/sbezverk/nx-telemetry-proto/tree/mod_update. You could use a `replace` statement in `go.mod` to use this version until things get merged. I don't think copying the `.proto` files or the generated code here will be accepted. @reimda correct me if I'm wrong here.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:23]: It would be nice to keep a space (newline) between 3rd party and telegraf packages .

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go
Comment (srebhan) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:66]: Can you please add a comment here?
```suggestion
	// Though unused in the code, required by protoc-gen-go-rpc to maintain (forward-)compatibility
	dialout.UnimplementedGRPCMdtDialoutServer
```

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (srebhan) [inline plugins/inputs/riemann_listener/riemann_listener.go:26]: Can you please either fork that repo or submit a PR upstream to update the generated protobuf?

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (srebhan) [inline plugins/inputs/riemann_listener/riemann_listener.go:207]: Is there any reason why you need a copy of the received events? Can't you directly iterate on the events?

**Review by srebhan — CHANGES_REQUESTED** [review]
Nice update @DStrand1! I have some more comments and a request to also get rid of the `riemann` stuff which also fails in CI. Can you please fix this upstream or in a fork of the upstream repo? I really want to avoid to open this box...

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (DStrand1) [reply plugins/inputs/riemann_listener/riemann_listener.go:26]: I have done a `replace` to use a fork of the `riemann-client-go` repo, and am going to submit a PR to that repo with the changes. I will PR to Telegraf if that PR is merged

File: plugins/inputs/riemann_listener/riemann_listener.go
Comment (DStrand1) [reply plugins/inputs/riemann_listener/riemann_listener.go:207]: We could directly iterate over the events, but since we are now using an updated fork of the `riemann` repo (i.e., not including our own copy of the `ProtocolBuffersToEvents()` function in Telegraf), I think it is cleaner to call this function in `riemanngo` instead of handling the logic from there ourselves

**Review by srebhan — APPROVED** [review]
Amazing work @DStrand1! Code looks good to me and I keep my fingers crossed for you to get your PRs accepted upstream.

**Review by sspaink — APPROVED** [review]

**Comment by sspaink** [comment]
Thanks for working on this, using the forked repository for now seems like a good idea to me. Just to make sure, @srebhan and @DStrand1 the intention isn't to wait for upstream to merge the changes before merging this pr?

**Comment by bai** [comment]
Not a reviewer but wanted to stop by and say thanks for fantastic work in this PR 🙏🏼

**Comment by srebhan** [comment]
@sspaink exactly. We can get rid of some pretty old libs with this PR and we shouldn't wait for upstream to get out of their holes... ;-)

**Pull request merged**
